From 3ce3e89b71e1eab157b5f5010e0d6f1f3b62d701 Mon Sep 17 00:00:00 2001
From: Stephano Cetola <scetola@reachtech.com>
Date: Tue, 29 Jul 2014 13:18:10 -0700
Subject: [PATCH 1/1] Fix for g2c kernel EV_VERSION

Signed-off-by: Stephano Cetola <scetola@reachtech.com>
---
 plugins/input-raw.c |    4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/plugins/input-raw.c b/plugins/input-raw.c
index 10454da..2efdfcb 100644
--- a/plugins/input-raw.c
+++ b/plugins/input-raw.c
@@ -50,6 +50,8 @@
 #define GRAB_EVENTS_WANTED	1
 #define GRAB_EVENTS_ACTIVE	2
 
+#define EV_VERSION_TSLIB  0x010000
+
 struct tslib_input {
 	struct tslib_module_info module;
 
@@ -83,7 +85,7 @@ static int check_fd(struct tslib_input *i)
 		return -1;
 	}
 
-	if (version < EV_VERSION) {
+	if (version < EV_VERSION_TSLIB) {
 		fprintf(stderr, "tslib: Selected device uses a different version of the event protocol than tslib was compiled for\n");
 		return -1;
 	}
-- 
1.7.10.4

