--- git/udev/udevd.c-orig	2013-08-06 11:26:49.540969681 -0700
+++ git/udev/udevd.c	2013-08-16 05:44:48.199413019 -0700
@@ -1223,7 +1223,7 @@ int main(int argc, char *argv[])
 			if (udev_set_run_path(udev, filename) == NULL)
 				goto exit;
 			mkdir(udev_get_run_path(udev), 0755);
-			err(udev, "error: runtime directory '%s' not writable, for now falling back to '%s'",
+			info(udev, "warning: runtime directory '%s' not writable, for now falling back to '%s'",
 			    udev_get_run_config_path(udev), udev_get_run_path(udev));
 		}
 	}
