From 11a1c736521856b82f1de8c7242aa114c52a066a Mon Sep 17 00:00:00 2001
From: Jeff Horn <outforaride@gmail.com>
Date: Thu, 17 Jan 2019 10:41:14 -0800
Subject: [PATCH] include/configs/g2h.h: Change dtb name for new test build

Signed-off-by: Jeff Horn <outforaride@gmail.com>
---
 include/configs/g2h.h | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/include/configs/g2h.h b/include/configs/g2h.h
index 2e945ff..c5dbf07 100644
--- a/include/configs/g2h.h
+++ b/include/configs/g2h.h
@@ -116,17 +116,17 @@
 
 /* kernel device tree options */
 #if defined(CONFIG_RD6_7P)
-#define CONFIG_DT "imx6dl-rd6-7p.dtb"
+#define CONFIG_DT "imx6dl-g2h-7p.dtb"
 #elif defined(CONFIG_RD6_7R)
-#define CONFIG_DT "imx6dl-rd6-7r.dtb"
+#define CONFIG_DT "imx6dl-g2h-7r.dtb"
 #elif defined(CONFIG_RD6_57R)
-#define CONFIG_DT "imx6dl-rd6-57r.dtb"
+#define CONFIG_DT "imx6dl-g2h-57r.dtb"
 #elif defined(CONFIG_RD6_57P)
-#define CONFIG_DT "imx6dl-rd6-57p.dtb"
+#define CONFIG_DT "imx6dl-g2h-57p.dtb"
 #elif defined(CONFIG_RD6_10P)
-#define CONFIG_DT "imx6dl-rd6-10p.dtb"
+#define CONFIG_DT "imx6dl-g2h-10p.dtb"
 #elif defined(CONFIG_RD6_10U)
-#define CONFIG_DT "imx6dl-rd6-10u.dtb"
+#define CONFIG_DT "imx6dl-g2h-10u.dtb"
 #else
 #define CONFIG_DT ""
 #endif
-- 
2.7.4

