diff -Naur qtmultimedia-opensource-src-5.1.1_original/src/gsttools/qgstreamervideowidget.cpp qtmultimedia-opensource-src-5.1.1/src/gsttools/qgstreamervideowidget.cpp
--- qtmultimedia-opensource-src-5.1.1_original/src/gsttools/qgstreamervideowidget.cpp	2013-12-25 19:34:38.623851754 -0800
+++ qtmultimedia-opensource-src-5.1.1/src/gsttools/qgstreamervideowidget.cpp	2013-12-25 20:24:54.411767300 -0800
@@ -124,21 +124,7 @@
     m_widget->installEventFilter(this);
     m_windowId = m_widget->winId();
 
-    m_videoSink = gst_element_factory_make ("xvimagesink", NULL);
-    if (m_videoSink) {
-        // Check if the xv sink is usable
-        if (gst_element_set_state(m_videoSink, GST_STATE_READY) != GST_STATE_CHANGE_SUCCESS) {
-            gst_object_unref(GST_OBJECT(m_videoSink));
-            m_videoSink = 0;
-        } else {
-            gst_element_set_state(m_videoSink, GST_STATE_NULL);
-
-            g_object_set(G_OBJECT(m_videoSink), "force-aspect-ratio", 1, (const char*)NULL);
-        }
-    }
-
-    if (!m_videoSink)
-        m_videoSink = gst_element_factory_make ("ximagesink", NULL);
+    m_videoSink = gst_element_factory_make ("glimagesink", NULL);
 
     gst_object_ref (GST_OBJECT (m_videoSink)); //Take ownership
     gst_object_sink (GST_OBJECT (m_videoSink));
diff -Naur qtmultimedia-opensource-src-5.1.1_original/src/gsttools/qgstreamervideowindow.cpp qtmultimedia-opensource-src-5.1.1/src/gsttools/qgstreamervideowindow.cpp
--- qtmultimedia-opensource-src-5.1.1_original/src/gsttools/qgstreamervideowindow.cpp	2013-12-25 19:34:38.623851754 -0800
+++ qtmultimedia-opensource-src-5.1.1/src/gsttools/qgstreamervideowindow.cpp	2013-12-25 19:36:01.171849442 -0800
@@ -68,7 +68,7 @@
     if (elementName)
         m_videoSink = gst_element_factory_make(elementName, NULL);
     else
-        m_videoSink = gst_element_factory_make("xvimagesink", NULL);
+        m_videoSink = gst_element_factory_make("glimagesink", NULL);
 
     if (m_videoSink) {
         gst_object_ref(GST_OBJECT(m_videoSink)); //Take ownership
